### ISSUE TITLE
CS2030 Finals Q&A 

### ISSUE THREAD

**Issue by seanlowjk** [nus-cs2030/2021-s2#252]
Hi All, 

We will be opening up a section for you to ask anything about Finals and PYP Questions with regards to CS2030. You can refer to https://libportal.nus.edu.sg/frontend/exam-papers to retrieve any PYPs for CS2030.  Do note that not all questions in the PYPs are relevant, so please filter them accordingly 

For the catch-up session, feel free to use the Finals Q&A Issue template to post any question with regards to that. Do take this opportunity to gain some class ;participation by answering and discussing the answers with your fellow peers. 

We will address any outstanding questions after Friday. 

Other than that, have a great reading week! 

Regards, 
CS2030 teaching team

**Comment by Guo-Yujia** [comment]
can tell some about the format of final exam, will we write on paper then  scan and submit? Will we have MCQ in the exam?

**Comment by Martini-Chwang** [comment]
Where can the answers for the PYP be found? O.o

**Comment by ryyant** [comment]
Hi, will the answer key be uploaded?

**Comment by JoelToh404** [comment]
any info on the format of the exam?

**Comment by ShyamalSankar** [comment]
Could you go through heap stack and meta space and how they would look like in a program run?

**Comment by itsjiaming** [comment]
Are we allowed to refer to our own notes or the internet for finals?

**Comment by KingRyo13** [comment]
Possibly any info on how in-depth Lazy eval, parallel and asynchronous topics will be tested?

**Comment by KevinTan1203** [comment]
Do you have any tips on what to study specifically for the upcoming finals?

**Comment by Joeltan15** [comment]
Could you re-explain stack and metaspace concepts again?

**Comment by salteefish** [comment]
Is google allowed for the exam?

**Comment by jessraphael** [comment]
Will sample answers be provided for the uploaded past year papers?

**Comment by arjann57** [comment]
Any info on the total number of marks for the paper?

**Comment by eugenelimnus** [comment]
Hi, would like to ask if we will be able to refer to Java 11 API during the exam?

**Comment by Deshun99** [comment]
Hi, what is the breakdown of the marks in finals?

**Comment by emfancypants** [comment]
hello, can the answer key to the past year papers be uploaded? thank you.

**Comment by nauyc** [comment]
hi, means our finals will be fully coding questions?

**Comment by elizabethh07** [comment]
can we use a monitor for finals?

**Comment by chamath2030** [comment]
Anyone able to advise how to tackle the following question

![image](https://user-images.githubusercontent.com/82226334/115586539-47ce6180-a2ff-11eb-8dc8-bef488fd9027.png)

**Comment by chamath2030** [comment]
what is a pure function? 

![image](https://user-images.githubusercontent.com/82226334/115587285-17d38e00-a300-11eb-8b2f-08dc9d76e500.png)

**Comment by jessraphael** [comment]
The return value of the function depends only on the input parameters passed to the function. The execution of the function has no side effects.

**Comment by salteefish** [comment]
Yep prof mentioned earlier in the announcement that Lec 12 content wont be tested (part of which entails pure functions) ^

**Comment by chamath2030** [comment]
thx @jessraphael and @salteefish .... sry didn't attend the last lecture :-(

**Comment by SW-bza** [comment]
Will the format be similar to PYPs like expected outputs for section 1?

**Comment by chamath2030** [comment]
hi guys, i understand that only the first part of lazy evaluation is tested. can i check if the following question is considered the first part? since we did learn about caching in the first part 

![image](https://user-images.githubusercontent.com/82226334/115653337-86950380-a361-11eb-9fcc-1a153b46eef2.png)

**Comment by ShyamalSankar** [comment]
When do we use fork, join and compute? Ie which context do we use it together with?

**Comment by wentinggy** [comment]
hi friends, I dont really understand what's going on in this question. Thank you!
![Screenshot 2021-04-23 at 9 54 10 PM](https://user-images.githubusercontent.com/77621160/115881551-70c93080-a47e-11eb-8815-75d2291f0bd3.png)

**Comment by JiaJun98** [comment]
Hello, I would like to ask what is the difference between thenApply() vs thenApplySync()? Does it affect the order of execution?

**Comment by lohwenxin** [comment]
![photo_2021-04-24_22-32-04](https://user-images.githubusercontent.com/80497924/115962238-ed2a4500-a54c-11eb-9bef-e5a377bcd4c9.jpg)

Hello! I would like to ask for this qn, if it is throwing a new exception, should the keyword be throw or throws? From the lecture slide the keyword was throw, so I am confused as the solution says throws.

**Comment by salteefish** [comment]
@JiaJun98 the Async keyword is used to determine if you want your thenApply() or thenApplyAsync() mapper function to run on the same or different thread respectively. Whether it affects the order of execution or not, i think it depends on the context of the question. For e.g. if the same thread is currently busy and you call a thenApply() mapper function, it will be executed at a later time compared to if you call a thenApplyAsync() function. On the other hand, if the same thread is not busy, then the order of execution for both thenApply() and thenApplyAsync() should be the same. Hope that helps!

**Comment by JiaJun98** [comment]
@salteefish Thank you for your help

**Comment by salteefish** [comment]
@lohwenxin i verified with my ta, it seems that there has been a typo with the provided ans key. The solution should be throw instead of throws

**Comment by JiaJun98** [comment]
@salteefish Thank you I got it

**Comment by Cattttw** [comment]
do we need to write java file in the final exam?

**Comment by downmeanhigh** [comment]
@lohwenxin nice question, was having the same concern

**Comment by SrinivasG99** [comment]
@Cattttw from what my TA tells me you need to write java files but it does not need to be compliable, with reference to last sems exam

**Comment by wenshienyap** [comment]
@wentinggy hellooo wentinggy! I think this question is just testing on our understanding of the map function and how to apply them in a pipeline way? so like from what i understand, option 3 will only work if x is positive. so if the initial x is negative, both bar() methods wont be applied on a. if x is positive but the first function applied makes x negative, the second one wont be applied. but not sure what you aren't sure about this question so if want to elaborate just go ahead!

**Comment by yowyuxuan** [comment]
Does anyone know about the exam format?

**Comment by zzikry98** [comment]
Hi, is RecursiveTask examinable?

**Comment by Archduck4** [comment]
Hi, is there any written information about the format of the paper?

**Comment by chamath2030** [comment]
all the best for later every1!

**Comment by ck-109** [comment]
Hi guys in case some of yall isn't aware, there's a link [](https://www.comp.nus.edu.sg/~cs2030/exam/) to some additional Q and A. Regarding exam format, (Short ans is no, we won't know the format till we see the paper...) All the best!!
